### ISSUE TITLE
Unexpected behavior of StringBuffer object within EUDPlayerLoop

### ISSUE THREAD

**Issue by Chromowolf** [armoha/euddraft#5]
```js
import tempcustomText as tct;
function afterTriggerExec() {
    var c = 0;
    const s = StringBuffer(255);
    // Actually I don't know the difference between insert and insertf,
    // not sure when to use which
    s.insertf(0, ""); // It seems that I cannot use append if I don't insert something first.
    EUDPlayerLoop()();
        s.append("a");
        c++;
    EUDEndPlayerLoop();
    s.Display();
    tct.print(c);
}
````
In the case of having 5 players in total, I'm expecting the following output on the screen:
```
aaaaa
5
```
but actually I got:
```
a
5
```
The above is just a simplified example. What I was trying to do is getting all the PNames on one line. 
Could you please explain why the StringBuffer object only got 1 "a" appended? not 5?

**Comment by Chromowolf** [comment]
It seems to me that every player has his own StringBuffer. The methods append(..) and insert(..) only apply to the current player's StringBuffer, but have no effect on other players':
```js
function onPluginStart() {
    //Suppose my playerID is 0
    setcurpl(2);
    const s = StringBuffer(255);
    setcurpl(1);
    s.insertf(0, "");
    s.append("a");
    setcurpl(0);
    s.insertf(0, "");
    s.append("b");
    s.Display();
}
```
Then in the game i got the message `b`
Why does it work like this?

**Comment by armoha** [comment]
It's intended feature. Like `CenterView`, `DisplayText`, `MinimapPing` or other desync actions, all methods of `StringBuffer` have built-in CurrentPlayer check functionality for convenience of end users.

`EUDPlayerLoop` not only repeats the block by count of active players, but also sets CurrentPlayer to corresponding player. You should use `EUDLoopPlayer` when you want to iterate players who meet the conditions and don't want to modify CurrentPlayer.

So following code will display `a` as many as active human players.
```js
function afterTriggerExec() {
    var c = 0;
    const s = StringBuffer(255);
    setcurpl(getuserplayerid());  // Set CurrentPlayer to own local id, so StringBuffer functions will execute to all players.
    s.insert(0);
    foreach(p: EUDLoopPlayer()) {
        s.append("a");
        c++;
    }
    s.Display();
    simpleprint(c);
}
```
If you don't like to set CurrentPlayer to desync value, here is alternative.
```js
function afterTriggerExec() {
    var c = 0;
    const s = StringBuffer(255);
    EUDPlayerLoop()();
        s.insert(0);
        foreach(p: EUDLoopPlayer()) {
            s.append("a");
        }
        s.Display();
        c++;
    EUDEndPlayerLoop();
    simpleprint(c);
}
```

Below is workflow of wrong code.
```js
function afterTriggerExec() {
    var c = 0;
    const s = StringBuffer(255);
    s.insert(0);  // this line won't work to players not equal to CurrentPlayer.
    EUDPlayerLoop()();  // Loop active players from P1 to P8 and set CurrentPlayer.
        s.append("a");  // This only run once when CurrentPlayer == local player id
        c++;
    EUDEndPlayerLoop();
    // CurrentPlayer value restores to previous value.
    s.Display();  // Again, it displays to CurrentPlayer.
    simpleprint(c);
}
```

**Comment by armoha** [comment]
It's mainly for allowing to write performant code without knowing all details of internal implementation.

Long time ago when `StringBuffer` is first introduced, there is no CurrentPlayer validation feature. Here is example code.
```js
function test() {
    const s = StringBuffer();
    EUDPlayerLoop()();
        s.print("bunch", "of", "text", "etc");
    EUDEndPlayerLoop();
}
```
In the past, this code wrote same text content at most 8 times and displayed text once to each player, because text writing functions were sync actions but text display function is desync action due to characteristic of `DisplayText`.

So to write performant code in the past, users should write as follows. (This is not the case in today.)
```js
function test() {
    const s = StringBuffer();
    EUDPlayerLoop()();
        if(getcurpl() == getuserplayerid()) {  // desync condition, you can replace it to condition IsUserCP() but it's added in recent version.
            s.insert(0);
            s.append("bunch", "of", "text", "etc");
        }
        s.Display();
    EUDEndPlayerLoop();
}
```
This was quite headache to users who are not familiar to desync things, so I made a change to add internal CurrentPlayer validation to all methods of `StringBuffer`. I hope this explanation is helpful to understanding how `StringBuffer` works.

**Comment by armoha** [comment]
> It seems to me that every player has his own StringBuffer.

Correct understanding is that method of `StringBuffer` is no-op when CurrentPlayer is not equal to local player id.

**Comment by Chromowolf** [comment]
Very informative, thx!
This makes me start to think of desync issues. The following is my understanding about the internal mechanism of how triggers modify memory. I don't know whether it is correct or not. Please kindly correct me if you spot any mistake:
The game data are stored in the memory for each player. There are public memory and private memory. The public memory stores the mutual data for each players, like each player's resource amounts, the existing units in the map and etc. (this part of memory should be the same for all players).  The private memory stores the individual data, like the user's player ID, the text being displayed for that player and etc. (this part of memory can be different for each player).
The game will desync as soon as there is any mismatch of the public data for each player. Am I right? 
Now suppose Starcraft only takes up 8 bytes of memory for each player. The first 4 bytes store public data, and the remaining 4 bytes store private data. They are:
```js
/*
Address:
(publ) 0x0: Ore amount of Player 0
(publ) 0x1: Ore amount of Player 1
(publ) 0x2: Trigger current player (this should be public data, right?)
(publ) 0x3: Other public data
(priv) 0x4: The current user's player ID
(priv) 0x5 and 0x6: A 2-byte string to be displayed to current user
(priv) 0x7: 0x00 (EOS) 
*/
```
Suppose we have 2 players now, and their initial memory is like following:
```js
/*
Address:               0     1     2     3     4     5     6     7
data for player 0:    0x05  0x0A  0x01  0xAB  0x00  'Y'   's'  0x00
data for player 1:    0x05  0x0A  0x01  0xAB  0x01  'N'   'o'  0x00
*/
```
Now we have the following eps: (some are pseudo code)
```js
import tempCustomText as tct.
function onPluginStart() {
    setcurpl(0);
    if ("player 0 has 5 ore") {
        setcurpl(1);
        action("Add 1 ore for current player"); 
    }
    DisplayText("ab");
    bwrite(0x3, 0xCD);
    if (tct.ptr(0x5) == "ab") {   //Here comes another question: how can we compare two strings?
        bwrite(0x0, 0x0E);
    }
}
```
Any expression in the code that results in any value change (or potentially change) for any player's memory, can be regarded as a trigger. (It seems that every expression is a trigger......)
**My basic understanding is that, for any single trigger, Starcraft will loop each player's memory. And for each player, Starcraft will check if the conditions are fulfilled based on the corresponding player's memory, and modify only the corresponding player's memory (if conditions are met).**
Please correct me if I'm wrong.
For example, the 1st expression: `setcurpl(0);` . The condition of this "trigger" is always. And Starcraft will first attach player 0's memory and modify the value in the address 0x2 to 0x00, and then goes into player 1's memory and modify the value in his address 0x2 to 0x00 also.
For the `if` statement, Starcraft will first go to player 0's memory to check whether address 0x0 has value 0x05, (in this case, yes), 
and then modify the value in 0x2 to 0x01 for player0, 
and add modify the value in 0x1 to 0x0B for player0,
now player 0's memory becomes: 
0x05  0x0B  0x01  0xAB  0x00  'Y'   's'  0x00
After this, Starcraft will go to player 1's memory and do the same thing, resulting in:
player 1's memory becomes: 
0x05  0x0B  0x01  0xAB  0x01  'N'   'o'  0x00.
Next, the statement `DisplayText("ab");` seems to be equivalent to
```js
if(bread(0x2) == bread(0x4))
    SetMemory(0x5, "ab"); //pseudo code
``` 
Starcraft goes into player 0's memory: check, condition not met, do nothing to player 0's memory
Starcraft goes into player 1's memory: check, condition met, change player 1's memory
Also, the statement `bwrite(0x3, 0xCD);` changes **both player's** memory:
Starcraft goes into player 0's memory: change the value in 0x3 to 0xCD
Starcraft goes into player 1's memory: change the value in 0x3 to 0xCD
Now we have:
```js
/*
Address:               0     1     2     3     4     5     6     7
data for player 0:    0x05  0x0B  0x01  0xCD  0x00  'Y'   's'  0x00
data for player 1:    0x05  0x0B  0x01  0xCD  0x01  'a'   'b'  0x00
*/
```
The last `if` statement goes the same procedure:
Starcraft goes into player 0's memory: the value is not "ab", do nothing.
Starcraft goes into player 1's memory: condition met, change the value at address 0x0 to 0x0E
It results in:
```js
/*
Address:               0     1     2     3     4     5     6     7
data for player 0:    0x05  0x0B  0x01  0xCD  0x00  'Y'   's'  0x00
data for player 1:    0x0E  0x0B  0x01  0xCD  0x01  'a'   'b'  0x00
*/
```
Now the public data mismatch, game desyncs.
So the reason the game desyncs is that using a **individual data** in private memory as **condition** and **change** the **mutual data** in public memory.
If my understanding is correct, then what should I do to achieve "cheating code" like "show me the money" in multi-player games? That is, to detect the players' chat message (I think this should be private memory?), and to change players' resource amout (public memory).

**Comment by armoha** [comment]
There is strcmp or memcmp for string comparison.

**Comment by armoha** [comment]
MSQC can share local data to public.

**Comment by armoha** [comment]
> This makes me start to think of desync issues.

There will be no desync issue with StringBuffer unless you put function call which has synchronous side-effects like CreateUnit into argument of method of StringBuffer, which is pretty rare to write and avoidable.

Public and private are terms related to accessibility. We say sync/shared and desync/local memory for more appropriate term. It's not private memory since everyone can read or write the address, but each individual can have different value in the address. (There are unwritable memories but it's unrelated topic so I'll skip.) In common notation, we call the first player which evaluated to value 0 for player 1. So it's weird to say player 0.

For accurate understanding of sync/desync, I use term `user` rather than `player`. Player can be either human, computer, neutral, rescuable, etc so it's wrong to use player in strict view.

There are no loop/attach/detach memory things.
`setcurpl(0);` - Set memory address 0x6509B0 and values of cache variable and cache codition to 0. Value of CurrentPlayer is local data; it does not have to be same for every users. Surely it is instant desync if we run `setcurpl(getuserplayerid()); CreateUnits(1, 0, CurrentPlayer, "Anywhere");`, since owner of unit is must be sync.

> For the if statement, Starcraft will first go to player 0's memory to check whether address 0x0 has value 0x05, (in this case, yes),
and then modify the value in 0x2 to 0x01 for player0,
and add modify the value in 0x1 to 0x0B for player0

Wrong. The condition is Ok for everyone so value modifications happen for all users. Say this single trigger as example;
```py
Trigger(
    Accumulate(P1, AtLeast, 1, Ore),
    CreateUnit(1, "Zerg Drone", P1, "Anywhere")
)
```
When the condition is met, does this trigger only spawns P1 drone to user of player 1 and does not spawn unit to any others? The answer is no, it creates unit to everyone's StarCraft.

Depict DisplayText to single condition and SetMemory is oversimplified since it do lots of work, 0x640B58, 0x640B60~ 11 chat addresses and chat timers etc.

I made chat command example map in past for feedback to SC:R developer in the past. You can find it if you look around starcraft blizz forum.

**Comment by Chromowolf** [comment]
Thank you for your patient and detailed explanation. You are a nice person.
I got this:
https://us.forums.blizzard.com/en/starcraft/t/eud-string-modification-broken-in-1-23-1/645
I tried the two example maps, using the newest version 1.23.3.7146, they both work! Fantastic job.
(The chatMoney.scx isn't sided with source code)

To be honest I've been playing around the detection of chat message using the address 0x00640B60 for the whole day. I've been reading your ChangeLog and also the function list of eudplib. But the "string object" in eps is so frustrating that I got tons of errors when trying to compare them.
The following is a naive skeleton:
```js
var curline = dwread(0x00640B58); //The initial value of NextDisplayTextLine is 10
var answerDetected = False;
function afterTriggerExec() {
    setcurpl(getuserplayerid()); //Suppose I'm in a 1-player game in multiplayer mode
	
    const answer = StringBuffer(256); //256 is enough? Max is 218
    answer.insert(0);
    answer.append(PName(getcurpl()), ": abc"); //Preset a string to detect whether I send the message "abc"
	
    //Suppose at most 1 new message is displayed between any two trigger cycles (to make it simple)
    if (curline != dwread(0x00640B58)) { //Some new message(s) have been displayed
        const chat = StringBuffer(256); //256 is enough? Max is 218
        chat.insert(0);
        chat.append(tct.str(0x00640B60 + 218 * curline));
        if (chat.length() > 0) {  //To be safe
            if (**"the content of the string [chat] is exactly the same as [answer]"**) {
                DoSomeActions...;
            }
        }
        curline = dwread(0x00640B58);
        **chat.delete(chat.epd, chat.length())**; //I wanna free the StringBuffer but i don't know how...
        //Maybe add a new method for object StringBuffer which simply free the whole string? 
    }
}
```
I tried the function `strcmp(s1, s2)` but the function list of eudplib didn't elaborate its usage. I don't know what type the argument should be. When I tried some, the game crashed or there were compiling errors.
Could you please give me more hint on this?

**Comment by armoha** [comment]
You don't need to use `tempCustomText` at all, unless legacy codes; all its functionalities are merged to basic support.
```js
// ↓ chat.append(tct.str(0x00640B60 + 218 * curline));
chat.append(ptr2s(0x00640B60 + 218 * curline));
```

**StringBuffer**.*pos* is last modified position (EPD), or you can use `GetMapStringAddr(s.StringIndex)`. But `StringBuffer` is mainly for print, that's why all StringBuffer has its own string, which can be used directly by string usage; unit name or trigger actions, etc (DisplayText, Transmission, Mission objective, ...)

You can't free StringBuffer; all StringBuffer has individual string (not just sequence of bytes, legal string in Map String section). StringBuffer is not EUDStruct. It's just python class so can't dynamically allocated, and `const answer = StringBuffer(256);` syntax is static initialization so you can't free. (You didn't read the link I attached in https://github.com/phu54321/euddraft/issues/18#issuecomment-590638712 :S)

**Comment by armoha** [comment]
Formatted string is recommended rather than wrappers like ptr2s or epd2s.
```js
// ↓ chat.append(tct.str(0x00640B60 + 218 * curline));
chat.appendf("{:s}", 0x00640B60 + 218 * curline);

**Comment by Chromowolf** [comment]
Thank you.
 I did read that link before I post the reply. Maybe I didn't fully understand that article. So I'll read it again and do some research about the concept "static variable". Just ignore that stupid "delete/free" statemet. Sorry for that.

Also thanks for the explanation about the sync/desync memory. I wanna make sure I understand it correctly:
```js
function afterTriggerExec() {
    //0x006CDDC4: Mouse Coordinate X  (4-byte local memory)
    //0x006CDDC8: Mouse Coordinate Y  (4-byte local memory)
    const mouseX = dwread(0x006CDDC4);
    if (mouseX < 100)        // "condition" is to read to local memory
        CreateUnits(1, 0, 0, "Anywhere");  //"Action" is to change shared memory (Create 1 Terran Marine for player 1)
}
```
Suppose there are 2 players in the game: player 1 (with playerID evaluated to 0) and player 2 (with playid 1). When the game start, player 1 moves his mouse to the left part of the screen (such that his own mouseX is less than 100), player 2 keeps his mouse still (player2's mouseX is greater than 100).
In the above case, the `if` condition turns out to be `True` for player 1, but `False` for player 2. Then, the memory modification is executed only for player 1's memory, but not player 2. So there is a mismatch of the two players' shared memory, hence the game goes desync. 
Is this the internal mechanism?
If I use the MSQC, I could prevent the game from being desync?

```js
function afterTriggerExec() {
    //0x006CDDC4: Mouse Coordinate X  (4-byte local memory)
    //0x006CDDC8: Mouse Coordinate Y  (4-byte local memory)
    const mouseX = dwread(0x006CDDC4);
    if (mouseX < 100)        // "condition" is to read to local memory
        simpleprint("mouseX<100");  //"Action" is to change local memory? 
}
```
In this case, suppose we also have the condition True for player 1, and False for player 2. But in this case the game won't be desync. And only player 1 sees the message "mouseX<100" but player 2 can't see any message displayed. This result confused me a little bit. I know the message to be displayed is a local memory. But isn't "simpleprint" meant to display the message for all players?

**Comment by armoha** [comment]
> But isn't "simpleprint" meant to display the message for all players?

Yeah, this is true in normal scenario. `simpleprint` sets `CurrentPlayer` to own player id before print, and restores `CurrentPlayer` to previous value. But with desync condition, since condition is evaluated true only for Player1, text shows to P1, and P2 can't see the text.

> If I use the MSQC, I could prevent the game from being desync?

Yes. You don't have to know underlying mechanism of MSQC to utilize it. But I can talk about it briefly. Suppose you click unit train button of building. Clicking button is naturally local event. According to button action, TrainUnit info `1F 00 ??` submits to QueueGameCommand, and it later shares to all players. (GameCommands: https://github.com/bwapi/bwapi/blob/master/bwapi/BWAPI/Source/BW/OrderTypes.h ) MSQC behaves following order; when you have any local data to send others, click invisible unit, send RightClick command, and restore previous selected units. Now you get shared data (invisible unit's waypoint coordinates), and MSQC converts them to sync Death Counter or Variable for use.

**Comment by Chromowolf** [comment]
Thx a lot! 
By following this link I learned how to write a key detection trigger in multiplayer game:
http://www.staredit.net/topic/17739/#3
(Just add [MSQC] and type some extra lines in my edd file to link a key to a unit's death)

But I still have no clue on how to link "(mouseX<100)" to "Death(cp, Exactly, .., ..)". I know I should do this, but don't know the syntax.
May I ask for the full manual for MSQC? I couldn't find it online...

**Comment by armoha** [comment]
> May I ask for the full manual for MSQC? I couldn't find it online...

https://cafe.naver.com/edac/78006
All manual of official plugins and library (Korean).

**Comment by Chromowolf** [comment]
Thanks a lot! Now I know how to utilize MSQC.
Btw, are you Artanis, the author of that manual?

**Comment by Chromowolf** [comment]
1 more thing about MSQC:
If I were to use EUDVariable in MSQC, and put the following code into my main.edd:
```js
[MSQC]
NotTyping; A: MyArray, 1
```
then, in which file / which function should I put `EUDRegisterObjectToNamespace ("MyArray", MyArray);` ? 
Where and how to define and initialize `MyArray` ?

**Comment by armoha** [comment]
> Btw, are you Artanis, the author of that manual?

Yeah, it's me.

>  in which file / which function should I put EUDRegisterObjectToNamespace ("MyArray", MyArray); ?
Where and how to define and initialize MyArray ?

```js
// eps
const send = EUDCreateVariables(32);
const receive = py_eval('[PVariable() for _ in range(32)]');
function onPluginStart() {
    foreach(i: py_range(32)) {
        DoActions(send[i].SetNumber(0xFFF0FFF));
        EUDRegisterObjectToNamespace(py_eval('"send{}".format(i)'), send[i]);
        EUDRegisterObjectToNamespace(py_eval('"receive{}".format(i)'), receive[i]);
    }
}
```

**Comment by Chromowolf** [comment]
(1)
> https://cafe.naver.com/edac/78006
All manual of official plugins and library (Korean).

I've read through the manual of MSQC and think there may be a typo? (I read the comment section and there is also someone pointing it out.)
I read the source code of MSQC.py, the `onInit()` defines how it parses the content in main.edd. 
Maybe the correct usage of "xy" would be like this when there is only one argument after the colon:
```
[non-shared condition 1]; xy, [variable or address] : [array or address]
```
For example:
```
NotTyping ; xy, 0x006CDDC4 : 18
```
This will result in copying the value in 0x006CDDC4 (mouse X coordinate) to the death count of unit 18 for current player.
What about when there are 2 values? I tried the following, it can compile but doesn't work:
```
NotTyping ; xy, 0x006CDDC4, 0x006CDDC8: 18, 24
```
What's the correct syntax?

Also there is another syntax that I found in MSQC.py:
```
[non-shared condition 1]; val, [variable or address] : [array or address]
```
Is it exactly the same as "xy" (with only 1 argument)?

-----------------------------------------

(2)
> // eps
const send = EUDCreateVariables(32);
...

Thank you. The above code generates 32 pairs of variables `send0, receive0, send1, receive1, ... , send31, receive31` for the main.edd to use under [MSQC],
also generates 32 pairs of variables `send[0], receive[0] ...` to use in main.eps. Is that right?
So I suppose `receive[0][0], receive[0][1], ... receive[0][7]` could take over what a DeathUnit can do.
Now I have:
main.edd:
```
...
[main.eps]
[MSQC]
NotTyping; xy, 0x006CDDC4 : receive0
[eudTurbo]
```
main.eps:
```js
const send = EUDCreateVariables(32);
const receive = py_eval('[PVariable() for _ in range(32)]');
function onPluginStart() {
    foreach(i: py_range(32)) {
        DoActions(send[i].SetNumber(0xFFF0FFF));
        EUDRegisterObjectToNamespace(py_eval('"send{}".format(i)'), send[i]);
        EUDRegisterObjectToNamespace(py_eval('"receive{}".format(i)'), receive[i]);
    }
}

function afterTriggerExec() {
    foreach(pID: EUDLoopPlayer()) {
        EUDPlayerLoop()();
            const mouse = StringBuffer(256);
            mouse.insert(0);
            mouse.appendf("{:c}{:n}\x04 mouse X: {}", pID, pID, receive[0][pID]);
            mouse.DisplayAt(pID);
        EUDEndPlayerLoop();
    }
}
```
So during the game, there should be messages printing each player's mouse's X coordinate. 
But when I was testing, everything is OK until I move my mouse to the left-most side of the screen: when X coordinate becomes 0, the value of receive[0][pID] suddenly becomes 4294967295 (0xFFFFFFFF). Is this a bug?

-----------------------------------------

(3)
If I change it to:
edd:
```js
...
NotTyping; xy, send0 : receive0
...
```
eps:
```js
...
DoActions(send[i].SetNumber(0x006CDDC4));
...
```
Then it won't work. I will see 4294967295 being printed. How to use the variable send0? What is the logic?

**Comment by armoha** [comment]
(1) https://cafe.naver.com/edac/91741
tldr; use `val` syntax rather than `xy` for sharing a value. `val` is made later to improve `xy`.

> For example:
> NotTyping ; xy, 0x6CDDC4 : Alan Turret
> This will result in copying the value in 0x6CDDC4 (mouse X coordinate) to the death count of Alan Turret for current player.

Importance of condition is discussed in #17 (comment).

With `xy` sytax, `[MSQC]` treats the source `0x6CDDC4` as `Position` type.
`Position` type means it stores position value in dword as **x + y * 65536**.
(e.g. CUnit::position {0x28}, CUnit::target_position {0x58}.)

Let's say the map width is 256 grids. When value in source is greater than 4095 (0xFFF), greastest value representable with map X coordinates *except most significant bit*, those values not masked with 0xFFF are ignored.
Well, `0x6CDDC4`,  relative mouse X coordinate from screen's left-upper corner is always less than 4095 so you won't have any problem.

But when sharing Screen Coordinate X as `condition ; xy, 0x628448 : Alan Turret` and your screen left border is at 7200 pixel (225 grids),
(1) 7200 pixel is 0x00001C20 in hexadecimal.
(2) with `xy`, 0x00001C20 is treated as x coordinate of `0xC20`, y coordinate of 0.
(3) CurrentPlayer's Alan Turret deaths is increased by 3104 (0xC20).

`val` syntax has no notion of `Position` type so always use `val` rather than `xy`.


(2)
> NotTyping; xy, 0x006CDDC4 : receive0

You need to write as `condition; xy, send0: receive0`
To distinguish between not sending value at all and sending value '0', return value is set to -1 when sending doesn't occur.

> So during the game, there should be messages printing each player's mouse's X coordinate.

Likely (not guaranteed) to print for only Player1, mouse's relative X coordinate of each human players.
You didn't specifiy CurrentPlayer at all, so it is unsure which player CurrentPlayer is. It can be any player since any trigger in any plugins can edit CurrentPlayer beforehand.

(3)
> DoActions(send[i].SetNumber(0x006CDDC4));
```js
var v;
v = 1;
DoActions(v.SetNumber(1));  // same semantic
v = dwread(0x6CDDC4);  // what you really meant
```

**Comment by Chromowolf** [comment]
Thank you for the explanation. Now I know the difference between `val` and `xy`. They are just using different bitmasks.
(1)
I know 256*32 - 1 = 0x1FFF, so we have 0x1FFF1FFF for xy coordinates in any 256x256 map. But due to translation problem, I can't understand why "excluding 1 most significant bit for each of x and y". Could you elaborate a little bit?
(2)
Seems that MSQC still cannot distinguish "not sending value at all" and "sending value 0":
main.edd: (I'm gonna send only 0 or 1, so it doesn't matter if I use `val` or `xy`)
```js
[MSQC]
NotTyping; val, send : receive
```
main.eps:
```js
var send = 0;
const receive = PVariable();
function onPluginStart() {
	EUDRegisterObjectToNamespace("send", send);
	EUDRegisterObjectToNamespace("receive", receive);
}
function afterTriggerExec() {
	send = !send;
	//print(receive[cp]); //print codes are omitted
}
```
I got -1 and 1 alternating, instead of 0 and 1 alternating.
PS: This time I'm pretty sure I'm sending values. I tried:
```js
[MSQC]
send.Exactly(0); xy, send : receive
```
and still got -1

**Comment by armoha** [comment]
(1)
We can't use whole bits of waypoint coordinates because of unit size, map borders and default unit position. When we right click unit close to map border, resulting waypoint is pushed into map so unit size won't escape map area.

(2)
I'm going to test this next week, having an exam and appointments on this weekend. You could verify whether `[MSQC]` did *not* send anything when variable was 0 by seeing APM stop to spike.

**Comment by Chromowolf** [comment]
> You could verify whether [MSQC] did not send anything when variable was 0 by seeing APM stop to spike.

When send=0, MSQC is actually sending value. But the received value is -1.
I don't know whether this is an intended feature.

Take your time and hope u do well on the exam

**Comment by armoha** [comment]
I wrongly posted comments in https://github.com/armoha/euddraft/issues/17#issuecomment-809695695
Anyway thanks for reporting!
